File: src/main/java/com/jplearning/service/DiscussionService.java
package com.jplearning.service;

import com.jplearning.dto.request.CommentRequest;
import com.jplearning.dto.request.DiscussionRequest;
import com.jplearning.dto.response.CommentResponse;
import com.jplearning.dto.response.DiscussionResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface DiscussionService {
    /**
     * Create a new discussion
     * @param userId ID of the user creating the discussion
     * @param request Discussion details
     * @return Created discussion response
     */
    DiscussionResponse createDiscussion(Long userId, DiscussionRequest request);

    /**
     * Get a discussion by ID
     * @param discussionId ID of the discussion
     * @return Discussion response
     */
    DiscussionResponse getDiscussionById(Long discussionId);

    /**
     * Update a discussion
     * @param discussionId ID of the discussion to update
     * @param userId ID of the user updating the discussion
     * @param request Updated discussion details
     * @return Updated discussion response
     */
    DiscussionResponse updateDiscussion(Long discussionId, Long userId, DiscussionRequest request);

    /**
     * Delete a discussion
     * @param discussionId ID of the discussion to delete
     * @param userId ID of the user deleting the discussion
     */
    void deleteDiscussion(Long discussionId, Long userId);

    /**
     * Get all discussions for a lesson
     * @param lessonId ID of the lesson
     * @param pageable Pagination information
     * @return Page of discussion responses
     */
    Page<DiscussionResponse> getDiscussionsByLesson(Long lessonId, Pageable pageable);

    /**
     * Get all discussions created by a user
     * @param userId ID of the user
     * @param pageable Pagination information
     * @return Page of discussion responses
     */
    Page<DiscussionResponse> getDiscussionsByUser(Long userId, Pageable pageable);

    /**
     * Add a comment to a discussion
     * @param discussionId ID of the discussion
     * @param userId ID of the user adding the comment
     * @param request Comment details
     * @return Created comment response
     */
    CommentResponse addComment(Long discussionId, Long userId, CommentRequest request);

    /**
     * Get a comment by ID
     * @param commentId ID of the comment
     * @return Comment response
     */
    CommentResponse getCommentById(Long commentId);

    /**
     * Update a comment
     * @param commentId ID of the comment to update
     * @param userId ID of the user updating the comment
     * @param request Updated comment details
     * @return Updated comment response
     */
    CommentResponse updateComment(Long commentId, Long userId, CommentRequest request);

    /**
     * Delete a comment
     * @param commentId ID of the comment to delete
     * @param userId ID of the user deleting the comment
     */
    void deleteComment(Long commentId, Long userId);

    /**
     * Get all comments for a discussion
     * @param discussionId ID of the discussion
     * @return List of comment responses
     */
    List<CommentResponse> getCommentsByDiscussion(Long discussionId);

    /**
     * Get all replies to a comment
     * @param commentId ID of the parent comment
     * @return List of comment responses (replies)
     */
    List<CommentResponse> getRepliesByComment(Long commentId);
}

File: src/services/discussion-service.ts
import api from '@/lib/api';

export interface Discussion {
  id: number;
  title: string;
  content: string;
  lessonId: number;
  userId: number;
  userName: string;
  userAvatar?: string;
  createdAt: string;
  updatedAt: string;
  commentCount: number;
}

export interface Comment {
  id: number;
  content: string;
  discussionId: number;
  userId: number;
  userName: string;
  userAvatar?: string;
  parentId?: number;
  createdAt: string;
  updatedAt: string;
  replies?: Comment[];
}

export interface DiscussionRequest {
  title: string;
  content: string;
  lessonId: number;
}

export interface CommentRequest {
  content: string;
  parentId?: number;
}

export interface PaginatedDiscussions {
  content: Discussion[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
}

const DiscussionService = {
  /**
   * Lấy danh sách thảo luận cho một bài học
   * @param lessonId ID của bài học
   * @param page Số trang, bắt đầu từ 0
   * @param size Số lượng mục trên mỗi trang
   */
  getDiscussionsByLesson: async (
    lessonId: number,
    page = 0,
    size = 10
  ): Promise<PaginatedDiscussions> => {
    const response = await api.get<PaginatedDiscussions>(
      `/discussions/lesson/${lessonId}?page=${page}&size=${size}`
    );
    return response.data;
  },

  /**
   * Tạo một thảo luận mới
   * @param request Thông tin thảo luận
   */
  createDiscussion: async (request: DiscussionRequest): Promise<Discussion> => {
    const response = await api.post<Discussion>('/discussions', request);
    return response.data;
  },

  /**
   * Lấy chi tiết thảo luận theo ID
   * @param discussionId ID của thảo luận
   */
  getDiscussionById: async (discussionId: number): Promise<Discussion> => {
    const response = await api.get<Discussion>(`/discussions/${discussionId}`);
    return response.data;
  },

  /**
   * Cập nhật thảo luận
   * @param discussionId ID của thảo luận
   * @param request Thông tin cập nhật
   */
  updateDiscussion: async (discussionId: number, request: DiscussionRequest): Promise<Discussion> => {
    const response = await api.put<Discussion>(`/discussions/${discussionId}`, request);
    return response.data;
  },

  /**
   * Xóa thảo luận
   * @param discussionId ID của thảo luận
   */
  deleteDiscussion: async (discussionId: number): Promise<void> => {
    await api.delete(`/discussions/${discussionId}`);
  },

  /**
   * Lấy danh sách bình luận cho một thảo luận
   * @param discussionId ID của thảo luận
   */
  getCommentsByDiscussion: async (discussionId: number): Promise<Comment[]> => {
    const response = await api.get<Comment[]>(`/discussions/${discussionId}/comments`);
    return response.data;
  },

  /**
   * Thêm bình luận vào thảo luận
   * @param discussionId ID của thảo luận
   * @param request Thông tin bình luận
   */
  addComment: async (discussionId: number, request: CommentRequest): Promise<Comment> => {
    const response = await api.post<Comment>(`/discussions/${discussionId}/comments`, request);
    return response.data;
  },

  /**
   * Lấy chi tiết bình luận theo ID
   * @param commentId ID của bình luận
   */
  getCommentById: async (commentId: number): Promise<Comment> => {
    const response = await api.get<Comment>(`/discussions/comments/${commentId}`);
    return response.data;
  },

  /**
   * Cập nhật bình luận
   * @param commentId ID của bình luận
   * @param request Thông tin cập nhật
   */
  updateComment: async (commentId: number, request: CommentRequest): Promise<Comment> => {
    const response = await api.put<Comment>(`/discussions/comments/${commentId}`, request);
    return response.data;
  },

  /**
   * Xóa bình luận
   * @param commentId ID của bình luận
   */
  deleteComment: async (commentId: number): Promise<void> => {
    await api.delete(`/discussions/comments/${commentId}`);
  },

  /**
   * Lấy danh sách phản hồi cho một bình luận
   * @param commentId ID của bình luận
   */
  getRepliesByComment: async (commentId: number): Promise<Comment[]> => {
    const response = await api.get<Comment[]>(`/discussions/comments/${commentId}/replies`);
    return response.data;
  }
};

export default DiscussionService; 